docs/proposals/super-simple-discovery-api.md

```diff
@@ -0,0 +1,306 @@
+
+```
+kubeadm join --cluster-info-url="https://example/mycluster.json"
+```
+
+This is really a shorthand for someone doing something like (assuming we support stdin with `-`):
+
+```
+curl https://example.com/mycluster.json | kubeadm join --cluster-info-file=-
+```
+
+If the user requires some auth to the HTTPS server (to keep the ClusterInfo object private) that can be done in the curl command equivalent.  Or we could eventually add it to `kubeadm` directly.
+
+### Method: JWS Token
+
+There won’t always be a trusted external endpoint to talk to and transmitting the locator file out of band is a pain.  However, we want something a bit more secure than just hitting HTTP and trusting whatever we get back.  In this case, we assume we have the following:
+  * An address for at least one of the API servers (which will implement this API).
+    * This address is technically an HTTPS URL base but is often expressed as a bare domain or IP.
+  * A shared secret token
+
+An interesting aspect here is that this information is often easily obtained before the API server is configured or started.  This makes some cluster bring-up scenarios much easier.
+
+We put some structure into the token.  It has both a token identifier and the token value, separated by a dot.  Example:
+
+```
+A81E5d4DwI.0ok9tB1QhB
+```
+
+The first part of the token is the `token-id`.  The second part is the `token-secret`.
+
+The user experience for joining a cluster would be something like:
+
+```
+kubeadm join --token=A81E5d4DwI.0ok9tB1QhB <address>
+```
+
+**Note:** This is logically a different use of the token from TLS bootstrap.  We should look to harmonize these usages and allow the same token to play double duty.  It is likely that the bare `--token` command line flag will indicate the token be used for both ClusterInfo and TLS bootstrap.
+
+#### Quick Primer on JWS
+
+[JSON Web Signatures](https://tools.ietf.org/html/rfc7515) are a way to sign, serialize and verify a payload.  It supports both symmetric keys (aka shared secrets) along with asymmetric keys (aka public key infrastructure or key pairs).  The JWS is split in to 3 parts:
+1. a header about how it is signed
+2. the clear text payload
+3. the signature.
+
+There are a couple of different ways of encoding this data -- either as a JSON object or as a set of BASE64URL strings for including in headers or URL parameters.  In this case, we are using a shared secret and the HMAC-SHA256 signing algorithm and encoding it as a JSON object.  The popular JWT (JSON Web Tokens) specification is a type of JWS.
+
+#### ClusterInfo Request
+
+The API server serves up a simple plain old HTTP endpoint for this.  The client does a HTTP GET to:
+
+```
+http://<address>/api/v1alpha1/clusterinfo/?token-id=<token-id>
```

Comment:
I'm torn here because this really is a singleton.  I'm thinking that for federation the federation interface and API will have another way to enumerate the downstream clusters.

How about we go with a collection and have a well known member that clients can assume.

That would change this to `http://<address>/api/v1/clusterinfo/this?token-id=<token-id>`.  `this` could easily be `self`.

Initially this would be the only resource in this collection.  For federation we may have a way for clusters to keep track of and report on other clusters.  Those would be listed via id or perhaps some other friendly name.  We may want redirects/aliases here if we can figure out how to support that.

Thoughts?